hypothesis-python/tests/conjecture/common.py

```diff
@@ -110,28 +110,50 @@ def fresh_data(*, random=None, observer=None) -> ConjectureData:
     )
 
 
+def clamped_shrink_towards(kwargs):
+    v = kwargs["shrink_towards"]
+    if kwargs["min_value"] is not None:
+        v = max(kwargs["min_value"], v)
+    if kwargs["max_value"] is not None:
+        v = min(kwargs["max_value"], v)
+    return v
+
+
 @st.composite
-def draw_integer_kwargs(
+def integer_kwargs(
     draw,
     *,
-    use_min_value=True,
-    use_max_value=True,
-    use_shrink_towards=True,
-    use_weights=True,
+    use_min_value=None,
+    use_max_value=None,
+    use_shrink_towards=None,
+    use_weights=None,
     use_forced=False,
 ):
     min_value = None
     max_value = None
     shrink_towards = 0
     weights = None
 
+    if use_min_value is None:
+        use_min_value = draw(st.booleans())
+    if use_max_value is None:
+        use_max_value = draw(st.booleans())
+    use_shrink_towards = draw(st.booleans())
+    if use_weights is None:
+        use_weights = (
+            draw(st.booleans()) if (use_min_value and use_max_value) else False
+        )
+
     # this generation is complicated to deal with maintaining any combination of
     # the following invariants, depending on which parameters are passed:
     #
     # (1) min_value <= forced <= max_value
     # (2) max_value - min_value + 1 == len(weights)
     # (3) len(weights) <= 255
```

Comment:
Not action-relevant for this PR, but I'm reminded that we planned to make `weights=` accept a dictionary of int-values to weights, with sum <= 1 and the residual probability spread across the remaining values.  That's going to be pretty annoying to deal with, though it'd also enable neat tricks like upweighting special values (powers of two, int_max +/- 1, etc.) in addition to our current tricks with endpoints which would be on a more standard footing.

Reply:
yeah, I have a wip of this stashed somewhere. It was sufficiently annoying to get working that I'm kicking it a fair bit down the road, so I was happy to put in the effort to upgrading our weights generator here.